### ISSUE TITLE
[ActiveRecord] A call to `pluck` fails with `ActiveRecord::UnmodifiableRelation` error

### ISSUE THREAD

**Issue by davidpiegza** [rails/rails#53260]
[Recent changes](https://github.com/rails/rails/commit/b0a6fe4c14995ea7f3548accb46977adcf8da7db#diff-79b53b2602bf702bdd8ce677e096be6a6923a54236e17237c16068a510078683R1975) in `QueryMethods` cause an error when using `pluck([String])` on a loaded relation.

```ruby
posts = Post.all
posts.to_a

puts posts.pluck("posts.id") # Fails with: ActiveRecord::UnmodifiableRelation: ActiveRecord::UnmodifiableRelation
```

[This line](https://github.com/rails/rails/commit/b0a6fe4c14995ea7f3548accb46977adcf8da7db#diff-79b53b2602bf702bdd8ce677e096be6a6923a54236e17237c16068a510078683R1976) sets the `references_values` in `QueryMethods#arel_column_with_table`:

```ruby
self.references_values |= [Arel.sql(table_name, retryable: true)]
```
but this fails with an `ActiveRecord::UnmodifiableRelation` error when the relation is already loaded ([source](https://github.com/rails/rails/blob/b0a6fe4c14995ea7f3548accb46977adcf8da7db/activerecord/lib/active_record/relation/query_methods.rb#L1728-L1730) where it throws the error).

Please note: It only fails when using `pluck` with a column name as string.

### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"
  
  # gem "rails"
  
  # If you want to test against edge Rails replace the previous line with this:
  gem "rails", github: "rails/rails", branch: "main"
  
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class BugTest < ActiveSupport::TestCase
  def test_association_stuff
    post = Post.create!
    
    posts = Post.all
    posts.to_a
    posts.pluck("posts.id")
    
    assert_equal 1, Post.count
  end
end
```

cc @kamipo as commit author of the changes.

**Comment by davidpiegza** [comment]
A possible fix is to add a `unless loaded?` check in [this line](https://github.com/rails/rails/commit/b0a6fe4c14995ea7f3548accb46977adcf8da7db#diff-79b53b2602bf702bdd8ce677e096be6a6923a54236e17237c16068a510078683R1976):

```ruby
self.references_values |= [Arel.sql(table_name, retryable: true)] unless loaded?
```

**Comment by fatkodima** [comment]
You should load the records inside the block, not just return the association:
```ruby
    values = ApplicationRecord.connected_to(role: :writing) do
      Post.pluck("posts.id")
    end
```

**Comment by davidpiegza** [comment]
@fatkodima Thanks for the quick response. This works for simple queries but it's tricky for complex queries when a base scope is returned and `pluck` is called outside of the `connected_to` block.

**Comment by fatkodima** [comment]
Building queries inside `connected_to` and running them outside and expecting them to run on the correct database is a programmer error.

**Comment by davidpiegza** [comment]
> expecting them to run on the correct database

I think it's not about the expectation on which database it runs. The issue is about "a call to `pluck` fails with an error on a loaded relation". 

For example, it also fails in this case (without using `connected_to`):
```ruby
posts = Post.all
posts.to_a
posts.pluck("posts.id")
```

I'll update the example above to simplify it.

**Comment by kamipo** [comment]
Thanks for the report. Fixed by #53264.
